cirq/contrib/quimb/mps_simulator.py

```diff
@@ -231,92 +235,243 @@ def sample(
+                delta_row = qubit1.row - qubit2.row
+                delta_col = qubit1.col - qubit2.col
+
+                if compare_row_first:
+                    if delta_row != 0:
+                        return delta_row
+                else:
+                    if delta_col != 0:
+                        return delta_col
+
+                return delta_row + delta_col
+            else:
+                # For non-2D, we just use the naive ordering.
+                return self.qubit_map[qubit1] - self.qubit_map[qubit2]
+
+        qubit_queue.sort(key=functools.cmp_to_key(qubit_ordering_fn))
+        return qubit_queue
+
+    def _sum_up(self, skip_tracing_out_for_qubits):
+        M = qtn.Tensor(1.0)
+
+        def _trace_out(i):
+            return skip_tracing_out_for_qubits and (i not in skip_tracing_out_for_qubits)
+
+        qubit_queue = self._get_qubit_queue()
+
+        while len(qubit_queue) > 0:
+            pop_idx = -1
+            # Compute the smallest badness:
+            for idx in range(len(qubit_queue)):
+                qubit = qubit_queue[idx]
+                # We define the badness as the number of elements the tensor M post
+                # aggregation and tracing out.
+                i = self.qubit_map[qubit]
+                new_inds = set(self.M[i].inds)
+                uncollapsed_inds = new_inds - set(M.inds)
+
+                badness = 1
+                for ind, shape in zip(self.M[i].inds, self.M[i].shape):
+                    if ind in uncollapsed_inds or not _trace_out(i):
+                        badness *= shape
+
+                if pop_idx == -1:
+                    smallest_badness = badness
+                    pop_idx = idx
+                elif smallest_badness > badness:
+                    smallest_badness = badness
+                    pop_idx = idx
+
+            # Pop the element to aggregate
+            qubit = qubit_queue.pop(pop_idx)
+
+            i = self.qubit_map[qubit]
+            Mi = self.M[i]
+            if _trace_out(i):
+                Mi = _sum_reduce(Mi, self.i_str(i))
+            M = _make_tensor_if_scalar(M @ Mi)
+        return M
+
+    def partial_state_vector(self, skip_tracing_out_for_qubits):
+        M = self._sum_up(
+            skip_tracing_out_for_qubits={
+                self.qubit_map[qubit] for qubit in skip_tracing_out_for_qubits
+            }
+        )
+        # Here, we rely on the formatting of the indices, and the fact that we have enough
+        # leading zeros so that 003 comes before 100.
+        sorted_ind = tuple(sorted(M.inds))
+        return M.fuse({'i': sorted_ind}).data
+
     def state_vector(self):
-        M = np.ones((1, 1))
-        for i in range(len(self.M)):
-            M = np.einsum('ni,npj->pij', M, self.M[i])
-            M = M.reshape(M.shape[0], -1)
-        assert M.shape[0] == 1
-        return M[0, :]
+        return self.partial_state_vector(skip_tracing_out_for_qubits={})
```

Comment:
Usually "tracing out" qubits means removing them. I.e., you trace out all qubits from a simulation you don't get the state vector, you get the number 1. Is this rather something like `skip_contracting_qubits={}`? or is it `trace_out_qubits={}`?

Reply:
the tensor network contraction simplification means I can now understand what's being controlled by this argument, although I have big questions about why you're using the bras you're using (see above). This isn't tracing out. It's "projecting out" or "selecting" the output states for these qubits

Reply:
You had the good intuition, and it was indeed very wrong. See complete comment above.